### DISCUSSION TITLE
Compare GoldenDict master, @xiaoyifang's fork and @vedgy's Qt WebEngine PR #1542

### DISCUSSION THREAD

**Discussion by vedgy** [goldendict/goldendict#1616] [Show and tell]
I am starting this discussion to save #1542 comments from further clutter.

My Reply to https://github.com/goldendict/goldendict/pull/1542#issuecomment-1509488914:
> > > the previous pull requests #1400 #1447
> > 
> > 
> > Have you ever looked at those diffs and the review comments? The reviews got nowhere, because @xiaoyifang couldn't even bother to reread and clean up the diff he proposed to merge.
> > My impressions stem from all the times I reviewed @xiaoyifang's commits in order to borrow something. I had to implement almost everything differently, because there were no clean, bug-free commits to borrow. It's clear that our opinions on the fork's quality differ substantially. This pull request's comments is not the place for bickering, which will achieve nothing. Any interested C++ developer can review the changes and make his/her own conclusions.
> 
> I have to see something , in #1447 ,I have changed 2000 more or less lines, in this PR, you have already changed 5000+ lines of code. ![image](https://user-images.githubusercontent.com/105986/232181204-47b0c501-edc7-428b-a386-c065332d807d.png)

My changes are carefully split into more than a hundred commits with detailed commit messages and source code comments, including `TODO (Qt WebEngine)` markers for unported or imperfectly ported features. Such neatly packaged changes can be reviewed easily enough commit by commit. They can be merged incrementally, because commits are atomic, self-sufficient and (almost) never break previously working functionality.

> You claim it has many bugs , why not list some of them for others to have a clear viewport.

Many bugs are pointed out in review comments to #1447. Later, I have also noticed careless commenting out of mutex locks and other thread synchronization code without replacement, bugs related to mouse click handling in `ArticleView`, wrong porting of deprecated Qt features that definitely breaks seldom-used GoldenDict features, to name a few. Some of those bugs have probably been fixed in later commits (no telling how robust the fix-ups are). But many likely remain. Plus, the track record suggests that many more bugs have almost certainly been introduced in the commits I haven't looked at.

One of the reasons for many breakages and poorly ported features in @xiaoyifang's fork is that he apparently never bothered to compile a Qt WebKit version of GoldenDict and see how it works and what behavior should be preserved.

In order to have any expectation of correctness, one would have to review all unreviewed changes in the fork and try to fix most introduced bugs. This is a very daunting task because of non-atomic commits, short and useless commit messages, few source code comments, frequent breakages, subsequent reverts and fix-ups - sometimes several months and tens/hundreds of commits after the original commit. I find it difficult to borrow something useful from the fork. Much more difficult still to feel confident enough to actually use the application compiled from such broken code.

-------------------
My reply to https://github.com/goldendict/goldendict/pull/1542#issuecomment-1508985862:
> A lot of other significant problems are also got addressed in the fork. When will they be fixed in this repo?

I am afraid the introduced bugs are more significant than the problems fixed. Plus some of the unreviewed fixes make the situation worse or at least have severe side effects. What is significant and important is highly subjective. I for one am happy enough with how GoldenDict master works. Merging in my open pull requests make it work almost perfectly for me.

> Can we get a release for classical WebKit version?

If you want a WebKit release, do something to help it materialize, e.g. contribute a CI that would make automated builds, without committing untrustable binaries compiled on your own system. But I am sure you don't need or want such a release and pose this question just to criticize others.

I personally don't need the release and am not prepared to put much work into it. A new tag would be nice, but the maintainer's unwillingness to create one shouldn't be a huge practical hurdle. I have proposed one workaround in https://github.com/goldendict/goldendict/issues/839#issuecomment-1013732316. Constructive discussion pertaining to a new release belongs to that issue.

**Comment by xiaoyifang** [comment]
> Many bugs are pointed out in review comments to #1447

I think they are just code guidelines .



> Plus, the track record suggests that many more bugs have almost certainly been introduced in the commits I haven't looked at.

based the fact on truth ,not on imagination. If you do not look at it or use it ,how can you tell about it.





> I have also noticed careless commenting out of mutex locks and other thread synchronization code without replacement

I did comment out QThread  and replaced it with QFuture

**Reply by nonwill** [reply]
> > Plus, the track record suggests that many more bugs have almost certainly been introduced in the commits I haven't looked at.
> 
> based the fact on truth ,not on imagination. If you do not look at it or use it ,how can you tell about it.

https://github.com/goldendict/goldendict/pull/1542#issuecomment-1509000232
![](https://user-images.githubusercontent.com/46510529/232122001-964b967e-dbd9-4ab0-abde-d7c37d77d40b.png)

Please, and please go back and handle the 💩.

**Reply by xiaoyifang** [reply]
This code is fine,I have not seen any problem here ,and it is not in the PRs we are currently talking.

**Comment by vedgy** [comment]
> > Many bugs are pointed out in review comments to #1447
> 
> I think they are just code guidelines .

@nonwill and I spotted and pointed out many obvious bugs in that review too.

> > Plus, the track record suggests that many more bugs have almost certainly been introduced in the commits I haven't looked at.
> 
> based the fact on truth ,not on imagination. If you do not look at it or use it ,how can you tell about it.

Whenever I look at your commits, I find mistakes and bugs, as well as poor coding/commit practices. Pointing out my expectations from the commits I haven't looked at is extrapolation, not idle speculation/imagination. What is the point of wasting my time to examine all of your commits and counting the average rate of bugs per commit? Careful review would likely take more time than it took you to churn out that mess.

> > I have also noticed careless commenting out of mutex locks and other thread synchronization code without replacement
> 
> I did comment out QThread and replaced it with QFuture

That was in later commits. Earlier commits often just comment out thread synchronization. An example I noticed recently while working on #1590: commenting out locks to fix-up a deadlock introduced (by you) many commits earlier via incorrect porting of a deprecated Qt feature, which consisted of making a recursive mutex non-recursive without giving it much thought. Needless to say, the end result is thread-unsafe.

**Reply by xiaoyifang** [reply]
> An example I noticed recently while working on #1590: commenting out locks to fix-up a deadlock introduced (by you) many commits earlier

the deadlock is not introuced by me ,it should easily  be reproduced in the original code by changing from RecursiveMutext to Mutex.
I changed the mutex and found the deadlock and comment out the extra locks.  because it is already locked by outer invoked method.

**Reply by xiaoyifang** [reply]
> > > Many bugs are pointed out in review comments to #1447
> > 
> > 
> > I think they are just code guidelines .
> 
> @nonwill and I spotted and pointed out many obvious bugs in that review too

plz give a real comment link about the bug

**Reply by vedgy** [reply]
> > An example I noticed recently while working on #1590: commenting out locks to fix-up a deadlock introduced (by you) many commits earlier
> 
> the deadlock is not introuced by me ,it should easily be reproduced in the original code by changing from RecursiveMutext to Mutex. I changed the mutex and found the deadlock and comment out the extra locks. because it is already locked by outer invoked method.

But why would anyone make the mutex nonrecursive without carefully analyzing why it has been made recursive? Only you could do such a careless change. Specifically, in 7b5511a7bd92fe04556faa7de044ff6d1b23d181 you did this:
```diff
-  Mutex() : QMutex( Recursive )
+  Mutex() : QMutex(  )
```

Then some time later another person fixed the deadlock by going back to recursive mutex in 889864a923977504ef79b4c90e112e26df9f4539:
```diff
-  Mutex() : QMutex(  )
+  Mutex() : QRecursiveMutex(  )
```
Several months later in 3d3088738ecc8541212649eb17f872cd6f9e38b4 you went back to nonrecursive mutex without any comment/reference as to why this was then safe.

Possibly relevant commits (hard to tell without explicit references or explanation):
* 206474364aa1bec99d4699b24bfb8ef8f7febd3b commented out mutex locks in mdx.cc;
* f65d47602abac444034c52453282184134ce8d26 added a mutex lock to `MdxDictionary::loadArticle()`;
* 4edd3b3987ff9126a4e30d26f363487e2b82e2ab added a couple more mutex locks.

> > > > Many bugs are pointed out in review comments to #1447
> > > 
> > > 
> > > I think they are just code guidelines .
> > 
> > 
> > @nonwill and I spotted and pointed out many obvious bugs in that review too
> 
> plz give a real comment link about the bug


https://github.com/goldendict/goldendict/pull/1447#discussion_r798464145
https://github.com/goldendict/goldendict/pull/1447#discussion_r795688178
https://github.com/goldendict/goldendict/pull/1447#discussion_r805999759

**Reply by nonwill** [reply]
So complicated an analysis is an excessive waste of time and energy, but it may not be able to impress xiaoyifang and also would not shake its **_stubbornness_**.

**Reply by xiaoyifang** [reply]
> [#1447 (comment)](https://github.com/goldendict/goldendict/pull/1447#discussion_r798464145)
> [#1447 (comment)](https://github.com/goldendict/goldendict/pull/1447#discussion_r795688178)
> [#1447 (comment)](https://github.com/goldendict/goldendict/pull/1447#discussion_r805999759)

They are during PR views ,haven't these been fixed? I think I follow the instruction and made changes.

**Reply by xiaoyifang** [reply]
> * [2064743](https://github.com/goldendict/goldendict/commit/206474364aa1bec99d4699b24bfb8ef8f7febd3b) commented out mutex locks in mdx.cc;
> * [f65d476](https://github.com/goldendict/goldendict/commit/f65d47602abac444034c52453282184134ce8d26) added a mutex lock to `MdxDictionary::loadArticle()`;
> * [4edd3b3](https://github.com/goldendict/goldendict/commit/4edd3b3987ff9126a4e30d26f363487e2b82e2ab) added a couple more mutex locks.

These 3 commits  ,  to be clear ,they are not in the PR #1400


this is an attempt to make fulltext creation parallel.   I removed the mutex and add a lock in the `getBlock` method to keep the critical area minimum at last.

Sometimes I may not get it right at first ,I have no excuse to share but the coding keeps improving

**Reply by vedgy** [reply]
> They are during PR views ,haven't these been fixed? I think I follow the instruction and made changes.

Some broken changes have been reverted; some workarounds replaced with less obviously broken workarounds; some bugs were never addressed (at least not in the PR branch). Note that these bugs have been found during cursory review. There were most likely plenty more. But the review process was so frustrating and time-consuming, mostly because of a host of unrelated changes and minor issues (which should never have been put into review in the first place), that I never had a chance to find them.

Anyone can check out the review comments and see what kind of changes were proposed for merging. Therefore, I am done discussing #1447, which is, by the way, a good representation of @xiaoyifang's development paradigm.

It's easy enough to find regressions introduced in @xiaoyifang's commits. But figuring out whether and where they have been subsequently fixed is very difficult and time-consuming.

> Sometimes I may not get it right at first ,I have no excuse to share but the coding keeps improving

The coding process may gradually improve with time and experience. The problem is in the careless development process:
1. make any, often "temporary" or experimental, change that appears to fix a perceived issue;
2. commit the change without considering possible consequences and forget about it;
3. revert, fix or rework in response to a user bug report (or just refuse to acknowledge and keep the reported regression as mentioned [here](https://github.com/goldendict/goldendict/pull/1542#issuecomment-1509081575)).

This way, regularly introduced regressions eventually get fixed in more-often used functionality. However, regressions accumulate in areas that are used less often or are less essential to most end users. Therefore, it's difficult to tell whether the overall shape of the application improves or degrades over time.

I personally don't want to experience and report/fix numerous regressions if most improvements and fixes in @xiaoyifang's version are not important to me. It's easier to implement important changes on top of a stable and reliable base. So I refuse to acknowledge the messy fork as a successor/replacement for GoldenDict master (this repository).

**Reply by xiaoyifang** [reply]
> 3\. revert, fix or rework in response to a user bug report (or just refuse to acknowledge and keep the reported regression as mentioned [here](https://github.com/goldendict/goldendict/pull/1542#issuecomment-1509081575)).

he mentioned but has not provided detail ,You do not verify and you claim I must be wrong.Such reckless.

@v-fox can you provided detail about this .
As far as I can remember ,  the configuration is not broken ,  the original code use dictionary name to compare dictionary ,I just use dictionaryId to keep the consistent and performance .

This is compatible issue to be far and intended.  You can not hold all the burdens in the old code and claim it is good.

https://github.com/xiaoyifang/goldendict/commit/c664ac315cb4bffe8bdd4b81713a0a5f74b98847
![image](https://user-images.githubusercontent.com/105986/232221948-95f028f9-15de-48a5-951a-e271699b6618.png)

**Reply by xiaoyifang** [reply]
> So I refuse to acknowledge the messy fork as a successor/replacement for GoldenDict master (this repository).

Let time tell

**Reply by v-fox** [reply]
> @v-fox can you provided detail about this . As far as I can remember , the configuration is not broken , the original code use dictionary name to compare dictionary ,I just use dictionaryId to keep the consistent and performance .
> This is compatible issue to be far and intended. You can not hold all the burdens in the old code and claim it is good.

As I've said in [the ignored report](https://github.com/xiaoyifang/goldendict/issues/47), the id numbers for dictionaries have changed and old groups are no longer populated due to unrecognized numbers and names don't seem to be used either. I don't care why it was done, only that re-creating groups by dragging a whole bunch of dictionaries in the UI is time-consuming and annoying (as writing the config by hand is impossible due to need for id generation), so I refuse to do it.

But, more importantly, it shows the attitude in 1) freely breaking any thing at any point 2) closing complains as 'by design'. Who knows what could be broken next just as nonchalantly.

**Comment by xiaoyifang** [comment]
> https://github.com/goldendict/goldendict/pull/1542#issuecomment-1509485122

@nonwill
though you keep talking about not borrow any idea from my repos.You just did it again.
https://github.com/xiaoyifang/goldendict/commit/223e67e503c0b5633ec5e8aa86692268ad22ea5b

the screenshot is taken from your blog https://autoptr.top/gdocr/GoldenDict-OCR-Changelog/
![image](https://user-images.githubusercontent.com/105986/232223777-8c5c4cab-0ca5-4b69-97b4-0c20bd7e5b85.png)

**Comment by nonwill** [comment]


**Reply by xiaoyifang** [reply]
all these nonsense. I'm not the person like you . 
  OCR is a feature I will not implemented ,the market already has too many of them and goldendict can work together with them .  Let OCR experts do their work. 

Borrow (Stolen) code without acknologment, blame others for all the fault. 
And always talk "sh*t here ,and there" ,while still benefit from it.

**Reply by nonwill** [reply]
> all these nonsense. I'm not the person like you .
> OCR is a feature I will not implemented ,the market already has too many of them and goldendict can work together with them . Let OCR experts do their work.

You have nosense to do it, and you can't ['steal' them](https://github.com/xiaoyifang/Capture2Text) either😄😄😄😄 😄😄😄😄  😄😄😄😄

> blame others for all the fault.

That really is you, 😄😄😄😄 😄😄😄😄

> And always talk "sh*t here ,and there"

That's for special you, 😄😄😄😄

**Comment by xiaoyifang** [comment]
@nonwill do not force me to make a statement page about all the features/ideas/code your have taken  from my fork and claim your own

**Reply by nonwill** [reply]
(https://github.com/goldendict/goldendict/pull/1542#issuecomment-1509000232)
Just do it as you like 哈哈哈哈哈哈哈哈哈

![handict_web_file111](https://user-images.githubusercontent.com/46510529/232263649-69c307be-7190-4918-8a78-299b499d303a.gif)

Don't say you are not envious, go back and copy your homework quickly.

And as a wise move,  it would be perfect and the best thing you do too to put this [Discussion](https://github.com/goldendict/goldendict/discussions/1616#discussion-5090031) link on your claim page.
